src/python/pants/util/objects.py

```diff
@@ -392,6 +500,167 @@ def satisfied_by_type(self, obj_type):
     return issubclass(obj_type, self._types)
 
 
+class DatatypeFieldDecl(namedtuple('DatatypeFieldDecl', [
+    'field_name',
+    'type_constraint',
+    'default_value',
+    'has_default_value',
+])):
+  """Description of a field, used in calls to datatype().
+
+  All elements of the list passed to datatype() are parsed into instances of this class by the
+  parse() classmethod.
+
+  `default_value` is only relevant if `has_default_value` is True.
+  """
+
+  class FieldDeclarationError(TypeError): pass
+
+  def __new__(cls, field_name, type_constraint=None, **kwargs):
+    """Parse the arguments into a type constraint and a default value.
+
+    If the argument `default_value` is provided (only by keyword), `has_default_value` is set to
+    True, and the argument `default_value` is used as this field's `default_value`. If
+    `default_value` is not provided (as when parsing a field declaration from a tuple), but
+    `type_constraint` is a TypeConstraint with the property `has_default_value` evaluating to True,
```

Comment:
Do any `TypeConstraint`s actually have default values? If not, let's remove the abstract method, property, and this bit of pydoc

Reply:
You're right! Will do.